test/pg_upgrade.bats

```
@@ -0,0 +1,117 @@
+    export GPUPGRADE_HOME="${STATE_DIR}/gpupgrade"
+    gpupgrade kill-services
+
+    # If this variable is set (to a master data directory), teardown() will call
+    # gpdeletesystem on this cluster.
+    NEW_CLUSTER=
+}
+
+teardown() {
+    skip_if_no_gpdb
+
+    gpupgrade kill-services
+    rm -r "$STATE_DIR"
+
+    if [ -n "$NEW_CLUSTER" ]; then
+        delete_cluster $NEW_CLUSTER
+    fi
+
+    gpstart -a
+
+    PSQL="$GPHOME"/bin/psql
+    $PSQL -d postgres -p $PGPORT -c "DROP TABLE IF EXISTS test_pg_upgrade CASCADE;"
+}
+
+# Takes an old datadir and echoes the expected new datadir path.
+upgrade_datadir() {
+    local base="$(basename $1)"
+    local dir="$(dirname $1)_upgrade"
+
+    # Sanity check.
+    [ -n "$base" ]
+    [ -n "$dir" ]
+
+    echo "$dir/$base"
+}
+
+
+# yes, this will fail once we allow an index on a partition table
+@test "pg_upgrade --check fails on a source cluster with an index on a partition table" {
+    skip_if_no_gpdb
+
+    # TODO: code factor this with execute.bats
+    PSQL="$GPHOME"/bin/psql
+
+    # Store the data directories for each source segment by port.
+    run $PSQL -AtF$'\t' -p $PGPORT postgres -c "SELECT port, datadir FROM gp_segment_configuration WHERE role = 'p'"
+    [ "$status" -eq 0 ] || fail "$output"
+
+    declare -a olddirs
+    while read -r port dir; do
+        olddirs[$port]="$dir"
+    done <<< "$output"
+
+    local masterdir="${olddirs[$PGPORT]}"
+    local newmasterdir="$(upgrade_datadir $masterdir)"
+
+    # add in a index on a partition table, which causes pg_upgrade --check to fail
+    $PSQL -d postgres -p $PGPORT -c "CREATE TABLE test_pg_upgrade(a int) DISTRIBUTED BY (a) PARTITION BY RANGE (a)(start (1) end(4) every(1));"
+    $PSQL -d postgres -p $PGPORT -c "CREATE UNIQUE INDEX fomo ON test_pg_upgrade (a);"
+
+    gpupgrade initialize \
+        --old-bindir "$GPHOME/bin" \
+        --new-bindir "$GPHOME/bin" \
+        --old-port "$PGPORT" \
+        --disk-free-ratio=0 3>&-
```

We ran into this one while fixing a different bug -- this call to `initialize` should fail, not succeed with a zero exit code. (See the ignored error propagation elsewhere in `InitializeSubStep()`.)

---

Yes. Now that we started to return an `err` in the one spot that we had missed, it fails as expected. So we prepend it with `run` so that test execution continues, and then check for an exit status of '1' and the logs for the expected error.